### PR TITLE
Issue #1200 - Make plain text thread dump downloadable

### PR DESCRIPTION
**Pull request by skosijer**
Spring Boot version 2.2.0.M5 released a new actuator endpoint that provides plain text thread dump.
This pull request makes the content of this endpoint downloadable for any service instance on JVM/Thread Dump page.

### CODE REVIEW

File: spring-boot-admin-server-ui/src/main/frontend/services/instance.js [hunk 1]
```diff
@@ -179,6 +179,19 @@ class Instance {
     return this.axios.get(uri`actuator/threaddump`);
   }
 
+  async downloadThreaddump(instanceName) {
+    return axios.get(uri`actuator/threaddump`, {
+      headers: {'Accept': 'text/plain'}
+    }).then( function (response) {
```
Comment (joshiste) [inline spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: please use async/await and not then

File: spring-boot-admin-server-ui/src/main/frontend/services/instance.js [hunk 2]
```diff
@@ -179,6 +179,19 @@ class Instance {
...
+      const fileSaver = require('file-saver');
+      const blob = new Blob([response.data], {type: 'text/plain;charset=utf-8'});
+      fileSaver.saveAs(blob, instanceName + '-threaddump.txt');
+    })
+      .catch(function (error) {
+        console.warn('Downloading threaddump failed: ', error);
```
Comment (joshiste) [inline spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: We should give some visible feedback if that fails, especially as there will be pre boot 2.2.0 applications which don't have support for this,

File: spring-boot-admin-server-ui/src/main/frontend/services/instance.js [hunk 3]
```diff
@@ -179,6 +179,19 @@ class Instance {
     return this.axios.get(uri`actuator/threaddump`);
   }
 
+  async downloadThreaddump(instanceName) {
+    return axios.get(uri`actuator/threaddump`, {
+      headers: {'Accept': 'text/plain'}
+    }).then( function (response) {
+      const fileSaver = require('file-saver');
```
Comment (joshiste) [inline spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: Please use a regular import.

File: spring-boot-admin-server-ui/src/main/frontend/services/instance.js [hunk 4]
```diff
@@ -179,6 +179,19 @@ class Instance {
     return this.axios.get(uri`actuator/threaddump`);
   }
 
+  async downloadThreaddump(instanceName) {
```
Comment (joshiste) [inline spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: There is no need to pass in the instance name. It is available as `this.registration.name`

File: spring-boot-admin-server-ui/src/main/frontend/services/instance.js [hunk 5]
```diff
@@ -179,6 +179,19 @@ class Instance {
...
+    return axios.get(uri`actuator/threaddump`, {
+      headers: {'Accept': 'text/plain'}
+    }).then( function (response) {
+      const fileSaver = require('file-saver');
+      const blob = new Blob([response.data], {type: 'text/plain;charset=utf-8'});
+      fileSaver.saveAs(blob, instanceName + '-threaddump.txt');
+    })
+      .catch(function (error) {
```
Comment (joshiste) [inline spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: please use try/catch in conjunction with async/await

File: spring-boot-admin-server-ui/src/main/frontend/services/instance.js (hunk 2 above)
Comment (skosijer) [reply spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: I completely agree with you. Would you suggest propagating the error from the service to an actual component and then using some messaging mechanism to display it there?

Thanks!

Comment (joshiste) [reply spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: I'd propose not to catch the exception here but in the `threaddump/index.vue` and act on the error.

Comment (skosijer) [reply spring-boot-admin-server-ui/src/main/frontend/services/instance.js]: Done.

**Comment by joshiste** [comment]
Merged with 2a637b47af81734b1bcfaae7bb9154d726c51cad. Thanks for your contribution!

**Pull request closed without merging**